extensions/kafka/src/main/java/io/deephaven/kafka/ingest/JsonNodeUtil.java

```diff
@@ -53,7 +52,10 @@ private static void checkNode(Object key, JsonNode node, boolean allowMissingKey
             throw new IllegalArgumentException(
                     String.format("Key '%s' not found in the record, and allowMissingKeys is false.", key));
         }
-        if (!allowNullValues && isNullOrMissingField(node)) {
+        // 'node==null || node.isMissingNode()' is OK here, because missing keys
+        // are allowed (which implicitly means null values for those keys are allowed).
+        // only explicit null values are disallowed.
+        if (!allowNullValues && node != null && node.isNull()) {
```

Comment:
the problem before was that `!allowMissingKeys && allowNullValues` didn't work — if you have a missing key, the `node` variable will just be `null` already (because we are calling this with the result of `node.get(key)`, which generally returns `null` if the key is missing.

But if the key **is** there, with an explicit null value, then you will have a non-null node reference (so `node != null)` but it will be a `com.fasterxml.jackson.databind.node.NullNode` (so `node.isNull()`)